modules/gmake/src/proj.lua

```diff
@@ -0,0 +1,677 @@
+			proj.defaultConfigurations,
+			proj.verbosity,
+			proj.phonies,
+			proj.shellType,
+			proj.defines,
+			proj.includeDirs,
+			proj.targetDir,
+			proj.intermediateDir,
+			proj.targetName,
+			proj.cppFlags,
+			proj.cFlags,
+			proj.cxxFlags,
+			proj.linkFlags,
+			proj.configurations,
+			proj.linkCmd,
+			proj.objects,
+			proj.allRule,
+			proj.targetRule,
+			proj.targetDirRule,
+			proj.objDirRule,
+			proj.cleanRule,
+			proj.prebuildRule,
+			proj.fileRules
+		}
+	end,
+	configurations = function (cfg)
+		return {
+			proj.configTargetDir,
+			proj.configIntermediateDir,
+			proj.configTargetName,
+			proj.configDefines,
+			proj.configCppFlags,
+			proj.configCFlags,
+			proj.configCxxFlags,
+			proj.configIncludeDirs,
+			proj.configLinkFlags,
+			proj.configBuildCommands
+		}
+	end
+}
+
+
+proj.HEADER_FILES = array.join(premake.C_HEADER_EXTENSIONS, premake.CXX_HEADER_EXTENSIONS, premake.OBJC_HEADER_EXTENSIONS)
+proj.SOURCE_FILES = array.join(premake.C_SOURCE_EXTENSIONS, premake.CXX_SOURCE_EXTENSIONS, premake.OBJC_SOURCE_EXTENSIONS)
+
+
+---
+-- Export the project's `.makefile` file.
+--
+-- @return
+--	True if the target `.makefile` file was updated; false otherwise.
+---
+function proj.export(prj)
+	export.eol('\r\n')
+	export.indentString('\t')
+	premake.export(prj, prj.exportPath, function()
+		premake.callArray(proj.elements.project, prj)
+	end)
+end
+
+
+---
+-- Prints the header at the top o fthe project file.
+--
+-- @param prj
+--  project to print header for
+---
+function proj.header(prj)
+	wl('# GNU Makefile project file autogenerated by Premake')
+	wl()
+end
+
+
+---
+-- Prints the default configuration information for the project.
+--
+-- @param prj
+--  project to print defaults for
+---
+function proj.defaultConfigurations(prj)
+	wl('ifndef config')
+	export.indent()
+	wl('config=' .. prj.configs[1].name:lower())
+	export.outdent()
+	wl('endif')
+	wl()
+end
+
+
+---
+-- Prints verbosity flags to the makefile.
+--
+-- @param prj
+--  project to print verbosity flags for
+---
+function proj.verbosity(prj)
+	wl('ifndef verbose')
+	export.indent()
+	wl('SILENT = @')
+	export.outdent()
+	wl('endif')
+	wl()
+end
+
+
+---
+-- Prints the shell type to the makefile
+--
+-- @param prj
+--  project to print shell type for
+---
+function proj.shellType(prj)
+	wl('SHELLTYPE := posix')
+	wl('ifeq (.exe,$(findstring .exe,$(ComSpec)))')
+	export.indent()
+	wl('SHELLTYPE := msdos')
+	export.outdent()
+	wl('endif')
+	wl()
+end
+
+
+---
+-- Prints the include directories for a project to the makefile.
+--
+-- @param prj
+--  project to print include directories for
+---
+function proj.includeDirs(prj)
+	local includeDirs = prj.includeDirs
+	if includeDirs ~= nil and #includeDirs > 0 then
+		local includeDirString = table.concat(table.map(includeDirs, function(key, value)
+			local relative = path.getRelative(prj.location, value)
+			return '-I' .. relative
+		end), ' ')
+
+		wl('INCLUDES = %s', includeDirString)
+	else
```

Comment:
If you wouldn't mind, could this be split up a little differently? I find the current version hard to read. Realize this might be a copy-paste thing.

```lua
local values = table.map(includeDirs, function(key, value)
   -- ...
end)

wl('INCLUDES = %s', table.concat(values, ' ')
```

Reply:
Split that up.  Would definitely agree it was a bit hard to read.